src/System.Text.Primitives/System/Text/Encoders/Utf8Decoder.cs

```diff
@@ -0,0 +1,264 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Primitives.System.Text.Encoders
+{
+    /// <summary>
+    /// Contains facilities for validating and extracting Unicode scalar values from UTF8 byte sequences.
+    /// </summary>
+    internal static class Utf8Decoder
+    {
+        // Assumption: bytesAvailable > 0
```

Comment:
The comment suggests that we have an assumption that bytesAvailable > 0 but then we check for == 0 in code. Update the comment to >= 0?